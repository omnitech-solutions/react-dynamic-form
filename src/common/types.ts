import { RJSFSchema, UiSchema, FormContextType } from '@rjsf/utils'

interface FormData {
  [key: string]: any
}

interface FormFieldError {
  property: string
  value: any
}

interface RJSFSchemasProps {
  schema?: RJSFSchema
  uiSchema?: UiSchema
  formContext?: FormContextType
  conditionalRules?: object
  formData?: FormData
  errors?: Array<FormFieldError>
}

interface KeyValueOption {
  key: string
  value: string
}

export {
  RJSFSchema,
  UiSchema,
  FormContextType,
  FormData,
  FormFieldError,
  RJSFSchemasProps,
  KeyValueOption
}
