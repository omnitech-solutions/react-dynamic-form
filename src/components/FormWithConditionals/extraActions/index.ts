import filterByMax from './filterByMax'
import uiOptionsTransform from './uiOptionsTransform'

const actions = {
  filterByMax,
  uiOptionsTransform
}

export default actions
