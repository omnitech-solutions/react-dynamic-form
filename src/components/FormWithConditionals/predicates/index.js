import { isMoment } from 'moment'
import predicate from 'predicate'
import isGreaterThanDaysBefore from './isGreaterThanDaysBeforePredicate'

predicate.moment = date => isMoment(date)
predicate.isGreaterThanDaysBefore = isGreaterThanDaysBefore
