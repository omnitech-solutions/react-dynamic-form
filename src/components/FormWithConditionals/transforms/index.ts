import dayjs from 'dayjs'

export default {
  tomorrow: () => dayjs().add(1, 'day'),
  iso: (date: string) => dayjs(date).toISOString()
}
