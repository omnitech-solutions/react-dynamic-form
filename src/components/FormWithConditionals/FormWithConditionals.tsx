// @ts-ignore
import Engine from 'json-rules-engine-simplified'

import { intersection, omit } from 'lodash'
import React, { Component } from 'react'
// @ts-ignore
import rulesRunner from 'rjsf-conditionals/lib/rulesRunner'
import { shouldRender, deepEquals } from '../../utils'
import { FormData, FormFieldError, FormContextType } from '../../common/types'
import { nonProdConsoleError } from '../../utils/debug-helpers'
import extraActions from './extraActions'

const ALLOWED_PROPS = ['formData', 'formContext', 'errors']
interface FormWithConditionalsState {
  formData: FormData
  errors: Array<FormFieldError>
  formContext: FormContextType
}

interface ShouldRenderOptions {
  keysToPick: Array<string>
}

type Props = {
  children?: React.ReactNode
  formData?: FormData
  errors?: Array<FormFieldError>
  formContext?: FormContextType
  metadata?: any
  onChange: any
  rules: any[]
  origUiSchema: any
  origSchema: any
}

/**
 * Used solely to limit the number of renders performed
 */
class FormWithConditionals extends Component<Props> {
  private formData: FormData
  private runRules: any

  private shouldUpdate: boolean

  constructor(props: Props) {
    super(props)

    this.handleChange = this.handleChange.bind(this)
    this.updateConf = this.updateConf.bind(this)

    const { formData = {}, origSchema, origUiSchema } = this.props
    this.formData = formData

    this.shouldUpdate = false
    this.state = { schema: origSchema, uiSchema: origUiSchema }

    this.runRules = rulesRunner(
      origSchema,
      origUiSchema,
      props.rules || [],
      Engine,
      extraActions
    )

    this.updateConf(formData)
  }

  UNSAFE_componentWillReceiveProps(
    nextProps: Readonly<Props>,
    _nextContext: any
  ) {
    const formDataChanged =
      nextProps.formData && !deepEquals(nextProps.formData, this.formData)

    if (formDataChanged) {
      this.updateConf(nextProps.formData)
      this.shouldUpdate = true
    } else {
      this.shouldUpdate =
        this.shouldUpdate ||
        !deepEquals(nextProps, { ...this.props, formData: nextProps.formData })
    }
  }

  // NOTE: DO NOT CHANGE UNLESS YOU WANT A VERY BAD DAY
  // This is used to ensure that we only re-render on change of form data
  // Changing this can introduce performance issues and bugs that are very difficult to fix
  shouldComponentUpdate(nextProps: Props) {
    const shouldRenderOpts: ShouldRenderOptions = {
      keysToPick: ALLOWED_PROPS
    }
    // @ts-ignore
    return shouldRender(nextProps, this.state, shouldRenderOpts)
  }

  // NOTE: DO NOT CHANGE UNLESS YOU WANT A VERY BAD DAY
  // This is used to ensure that we only re-render on change of form data
  // Changing this can introduce performance issues and bugs that are very difficult to fix
  static getDerivedStateFromProps(
    props: Props,
    state: FormWithConditionalsState
  ) {
    const allowedPropsNotFound =
      intersection(ALLOWED_PROPS, Object.keys(props)).length === 0
    if (allowedPropsNotFound) return null

    const shouldUpdateState = shouldRender(state, props, {
      keysToPick: ALLOWED_PROPS
    })
    if (!shouldUpdateState) return null

    const { formData, errors, formContext } = props

    return {
      formData,
      errors,
      formContext
    }
  }

  handleChange(change: any) {
    const { formData } = change
    const updTask = this.updateConf(formData)

    const { onChange } = this.props
    if (onChange) {
      updTask
        .then((conf: any) => {
          const confFormData = {
            ...omit(conf.formData, 'metadata')
          }
          const updChange = { ...change, ...conf, formData: confFormData }

          onChange(updChange)
          return conf
        })
        .catch((ex: any) => {
          nonProdConsoleError(ex)
        })
    }
  }

  updateConf(formData: FormData) {
    this.formData = formData
    return this.runRules({ ...formData, metadata: this.props.metadata }).then(
      (conf: { formData: any; schema: any; uiSchema: any }) => {
        const confFormDataCpy = omit(conf.formData, 'metadata')
        const dataChanged = !deepEquals(this.formData, confFormDataCpy)
        this.formData = confFormDataCpy

        const newState = { schema: conf.schema, uiSchema: conf.uiSchema }
        const confChanged = !deepEquals(newState, this.state)
        if (dataChanged || confChanged) {
          this.shouldUpdate = true
          this.setState(newState)
        }

        return conf
      }
    )
  }

  render() {
    const { children }: Props = this.props
    return <span>{children}</span>
  }
}

export default FormWithConditionals
