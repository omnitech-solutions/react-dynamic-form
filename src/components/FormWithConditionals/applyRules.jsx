/* eslint-disable react/prop-types */
// Copied directly from https://github.com/RxNT/rjsf-conditionals/blob/master/src/applyRules.js and then changes applied to suit our needs.

import Engine from 'json-rules-engine-simplified'
import PropTypes from 'prop-types'

import React, { useState } from 'react'

import { DEFAULT_ACTIONS } from 'rjsf-conditionals/lib/actions'
import validateAction from 'rjsf-conditionals/lib/actions/validateAction'
import { isDevelopment } from 'rjsf-conditionals/lib/utils'
import { deepEquals } from 'react-jsonschema-form/lib/utils'
import { nonProdConsoleError } from '../helpers/debug-helper'

import extraActions from './extraActions'
import FormWithConditionalsExt from './FormWithConditionalsExt'

const applyRules = (schema, uiSchema, rules) => {
  if (isDevelopment()) {
    const propTypes = {
      Engine: PropTypes.func.isRequired,
      rules: PropTypes.arrayOf(
        PropTypes.shape({
          conditions: PropTypes.object.isRequired,
          order: PropTypes.number,
          event: PropTypes.oneOfType([
            PropTypes.shape({
              type: PropTypes.string.isRequired
            }),
            PropTypes.arrayOf(
              PropTypes.shape({
                type: PropTypes.string.isRequired
              })
            )
          ])
        })
      ).isRequired,
      extraActions: PropTypes.object
    }

    PropTypes.checkPropTypes(
      propTypes,
      { rules, Engine, extraActions },
      'props',
      'react-jsonschema-form-manager'
    )

    rules
      .reduce((agg, { event }) => agg.concat(event), [])
      .forEach(({ type, params }) => {
        // Find associated action
        const action = extraActions[type]
          ? extraActions[type]
          : DEFAULT_ACTIONS[type]
        if (action === undefined) {
          nonProdConsoleError(`Rule contains invalid action "${type}"`)
          return
        }

        validateAction(action, params, schema, uiSchema)
      })
  }

  return FormComponent => props => {
    const metadata = {}

    const [currentSchema, setSchema] = useState(schema)
    const [currentUiSchema, setUiSchema] = useState(uiSchema)
    const [currentFormData, setSetFormData] = useState(props.formData)
    const [rerenderCount, setSetRenderCount] = useState(0)

    const nextState = {
      schema: props.schema,
      uiSchema: props.uiSchema,
      formData: props.formData
    }

    const prevState = {
      schema: currentSchema,
      uiSchema: currentUiSchema,
      formData: currentFormData
    }

    if (!deepEquals(prevState, nextState)) {
      setSchema(nextState.schema)
      setUiSchema(nextState.uiSchema)
      setSetFormData(nextState.formData)
      setSetRenderCount(rerenderCount + 1)
    }

    return (
      <FormWithConditionalsExt
        {...props}
        Form={FormComponent}
        origSchema={schema}
        origUiSchema={uiSchema}
        metadata={metadata.formData}
      />
    )
  }
}

export default applyRules
