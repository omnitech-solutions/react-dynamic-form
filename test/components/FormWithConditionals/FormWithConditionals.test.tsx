import React from 'react'
import sinon from 'sinon'
import { expect } from 'chai'
import {
  act,
  create,
  ReactTestInstance,
  ReactTestRenderer
} from 'react-test-renderer'
import FormWithConditionals from '../../../src/components/FormWithConditionals/FormWithConditionals'
import { RJSFSchemasProps } from '../../../src/common/types'

const props: RJSFSchemasProps = {
  schema: {
    type: 'object',
    required: ['firstName'],
    properties: {
      firstName: {
        type: 'string',
        title: 'First name'
      },
      lastName: {
        type: 'string',
        title: 'Last name'
      }
    }
  },
  uiSchema: {
    firstName: {
      'ui:autofocus': true,
      'ui:emptyValue': ''
    }
  },
  formData: { firstName: 'some-first-name', lastName: 'some-last-name' },
  errors: [{ property: 'firstName', value: 'some-first-name-error' }],
  formContext: { currencySymbol: '$' }
}

let subject: ReactTestInstance | null
let root: ReactTestRenderer | null

const sandbox = sinon.createSandbox()
let prototype: any = null

describe('FormWithConditionals', () => {
  const setPropsWithField = ({
    name,
    value
  }: {
    name: string
    value: string
  }) => {
    act(() => {
      // @ts-ignore
      root.update(
        <FormWithConditionals
          {...props}
          formData={{ ...props.formData, [name]: value }}
        />
      )
    })
  }

  beforeEach(() => {
    root = create(<FormWithConditionals {...props} />)
    subject = root.getInstance()

    prototype = FormWithConditionals.prototype
    sandbox.spy(prototype, 'shouldComponentUpdate')
  })

  afterEach(() => {
    sandbox.restore()
  })

  it('sets state from given props', () => {
    // @ts-ignore
    const { state } = subject

    expect(Object.keys(state)).to.eql([
      'schema',
      'uiSchema',
      'formData',
      'errors',
      'formContext'
    ])

    // // @ts-ignore
    const { formData, errors, formContext } = state

    expect(formData).to.eql(props.formData)
    expect(errors).to.eql(props.errors)
    expect(formContext).to.eql(props.formContext)
  })

  it('ignores unexpected props', () => {
    // make assertions on root
    // @ts-ignore
    expect(root.toJSON()).to.matchSnapshot()

    act(() => {
      // @ts-ignore
      root.update(<FormWithConditionals {...props} unknown={2} />)
    })

    // update with some different props
    // @ts-ignore
    expect(root.toJSON()).to.matchSnapshot()
    // sandbox

    expect(prototype.shouldComponentUpdate.calledOnce).to.be.true
    expect(prototype.shouldComponentUpdate.returned(false)).to.be.true
  })

  describe('formData prop', () => {
    context('with changes', () => {
      it('should render', () => {
        setPropsWithField({ name: 'firstName', value: 'some-other-first-name' })

        expect(prototype.shouldComponentUpdate.calledOnce).to.be.true
        expect(prototype.shouldComponentUpdate.returned(true)).to.be.true
      })

      context('with multiple changes', () => {
        it('re-renders twice', () => {
          setPropsWithField({
            name: 'firstName',
            value: 'some-other-first-name'
          })
          setPropsWithField({ name: 'firstName', value: 'another-first-name' })

          expect(prototype.shouldComponentUpdate.calledOnce).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(true)).to.be
            .true
          expect(prototype.shouldComponentUpdate.secondCall.returned(true)).to
            .be.true
        })
      })

      context('with duplicate changes', () => {
        it('skips rendering when no changes detected', () => {
          setPropsWithField({
            name: 'firstName',
            value: 'some-other-first-name'
          })
          setPropsWithField({
            name: 'firstName',
            value: 'some-other-first-name'
          })

          expect(prototype.shouldComponentUpdate.calledTwice).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(true)).to.be
            .true
          expect(prototype.shouldComponentUpdate.secondCall.returned(false)).to
            .be.true
        })
      })

      context('with another duplicate changes', () => {
        it('re-renders once', () => {
          // @ts-ignore
          setPropsWithField({})
          setPropsWithField({
            name: 'firstName',
            value: 'some-other-first-name'
          })

          expect(prototype.shouldComponentUpdate.calledTwice).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(false)).to
            .be.true
          expect(prototype.shouldComponentUpdate.secondCall.returned(true)).to
            .be.true
        })
      })
    })

    context('with no formData changes', () => {
      it('should not render', () => {
        // @ts-ignore
        setPropsWithField({ ...props.formData })

        expect(prototype.shouldComponentUpdate.calledOnce).to.be.true
        expect(prototype.shouldComponentUpdate.firstCall.returned(false)).to.be
          .true
      })
    })
  })

  describe('errors prop', () => {
    const getFieldErrorFor = ({
      name,
      message
    }: {
      name: string
      message: string
    }) => ({
      property: name,
      name: 'required',
      message,
      stack: `.${name} ${message}`,
      schemaPath: `#/fields/${name}/required`
    })

    const setPropsWithError = ({
      name,
      message
    }: {
      name: string
      message: string
    }) => {
      act(() => {
        // @ts-ignore
        root.update(
          <FormWithConditionals
            {...props}
            errors={[getFieldErrorFor({ name, message })]}
          />
        )
      })
    }

    context('with changes', () => {
      it('should render', () => {
        setPropsWithError({ name: 'firstName', message: 'some-field-error' })

        expect(prototype.shouldComponentUpdate.calledOnce).to.be.true
        expect(prototype.shouldComponentUpdate.firstCall.returned(true)).to.be
          .true
      })

      context('with multiple changes', () => {
        it('re-renders twice', () => {
          setPropsWithError({
            name: 'firstName',
            message: 'some-other-first-name-field-error'
          })

          setPropsWithError({
            name: 'firstName',
            message: 'another-first-name-field-error'
          })

          expect(prototype.shouldComponentUpdate.calledTwice).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(true)).to.be
            .true
          expect(prototype.shouldComponentUpdate.secondCall.returned(true)).to
            .be.true
        })
      })

      context('with duplicate changes', () => {
        it('skips rendering when no changes detected', () => {
          setPropsWithError({
            name: 'firstName',
            message: 'some-other-first-name-field-error'
          })

          setPropsWithError({
            name: 'firstName',
            message: 'some-other-first-name-field-error'
          })

          expect(prototype.shouldComponentUpdate.calledTwice).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(true)).to.be
            .true
          expect(prototype.shouldComponentUpdate.secondCall.returned(false)).to
            .be.true
        })
      })

      context('with another duplicate changes', () => {
        it('re-renders once', () => {
          // @ts-ignore
          setPropsWithField({})

          setPropsWithError({
            name: 'firstName',
            message: 'some-other-first-name-field-error'
          })

          expect(prototype.shouldComponentUpdate.calledTwice).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(false)).to
            .be.true
          expect(prototype.shouldComponentUpdate.secondCall.returned(true)).to
            .be.true
        })
      })
    })

    context('with no errors changes', () => {
      it('should not render', () => {
        // @ts-ignore
        setPropsWithField({})

        expect(prototype.shouldComponentUpdate.calledOnce).to.be.true
        expect(prototype.shouldComponentUpdate.firstCall.returned(false)).to.be
          .true
      })
    })
  })

  describe('formContext prop', () => {
    const setPropsWithFormContextChanges = ({
      currencySymbol
    }: {
      currencySymbol: string
    }) => {
      act(() => {
        // @ts-ignore
        root.update(
          <FormWithConditionals
            {...props}
            formContext={{ ...props.formContext, currencySymbol }}
          />
        )
      })
    }

    context('with changes', () => {
      it('should render', () => {
        setPropsWithFormContextChanges({ currencySymbol: '£' })

        expect(prototype.shouldComponentUpdate.calledOnce).to.be.true
        expect(prototype.shouldComponentUpdate.firstCall.returned(true)).to.be
          .true
      })

      context('with multiple changes', () => {
        it('re-renders twice', () => {
          setPropsWithFormContextChanges({
            currencySymbol: '£'
          })

          setPropsWithFormContextChanges({
            currencySymbol: '€'
          })

          expect(prototype.shouldComponentUpdate.calledTwice).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(true)).to.be
            .true
          expect(prototype.shouldComponentUpdate.secondCall.returned(true)).to
            .be.true
        })
      })

      context('with duplicate changes', () => {
        it('skips rendering when no changes detected', () => {
          setPropsWithFormContextChanges({
            currencySymbol: '£'
          })

          setPropsWithFormContextChanges({
            currencySymbol: '£'
          })

          expect(prototype.shouldComponentUpdate.calledTwice).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(true)).to.be
            .true
          expect(prototype.shouldComponentUpdate.secondCall.returned(false)).to
            .be.true
        })
      })

      context('with another duplicate changes', () => {
        it('re-renders once', () => {
          // @ts-ignore
          setPropsWithField({})

          setPropsWithFormContextChanges({
            currencySymbol: '£'
          })

          expect(prototype.shouldComponentUpdate.calledTwice).to.be.true
          expect(prototype.shouldComponentUpdate.firstCall.returned(false)).to
            .be.true
          expect(prototype.shouldComponentUpdate.secondCall.returned(true)).to
            .be.true
        })
      })
    })

    context('with no errors changes', () => {
      it('should not render', () => {
        // @ts-ignore
        setPropsWithField({})

        expect(prototype.shouldComponentUpdate.calledOnce).to.be.true
        expect(prototype.shouldComponentUpdate.firstCall.returned(false)).to.be
          .true
      })
    })
  })
})
