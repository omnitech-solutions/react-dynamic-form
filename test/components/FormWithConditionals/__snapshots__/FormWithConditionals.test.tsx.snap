// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormWithConditionals > ignores unexpected props 1`] = `<span />`;

exports[`FormWithConditionals > ignores unexpected props 2`] = `<span />`;
